### PR TITLE
Added FinalPath implementation

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

OutputModules placed on a FinalPath do not cause unscheduled execution. This ability was requested by HLT.

As part of work, consolidated all copied code used in various OutputModuleBase implementations into one class.

#### PR validation:

Code compiles. All framework related unit tests (included the newly added ones) pass.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Integration/test/test_finalpath.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+found thing 'endRun' TEST
+EOF
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --schedule --task >& finalpath.log || die "failed test_finalpath_cfg.py --schedule --task" $?
+grep "thing '.*' TEST" finalpath.log | diff finalpath_expected_not_found.log - || die "differences for test_finalpath_cfg.py --schedule --task" $?
+
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --endpath >& finalpath.log || die "failed test_finalpath_cfg.py --endpath" $?
+grep "thing '.*' TEST" finalpath.log | diff finalpath_expected_empty.log - || die "differences for test_finalpath_cfg.py --endpath" $?
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --endpath >& finalpath.log || die "failed test_finalpath_cfg.py --schedule --endpath" $?
```
Comment (makortel) [inline FWCore/Integration/test/test_finalpath.sh:38]: Should this have
```suggestion
cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --schedule --endpath >& finalpath.log || die "failed test_finalpath_cfg.py --schedule --endpath" $?
```
?

File: FWCore/Integration/test/test_finalpath.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+found thing '' TEST
+found thing 'beginLumi' TEST
+found thing 'endLumi' TEST
+found thing 'beginRun' TEST
+found thing 'endRun' TEST
+EOF
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --endpath --task >& finalpath.log || die "failed test_finalpath_cfg.py --endpath --task" $?
+grep "thing '.*' TEST" finalpath.log | diff finalpath_expected_found.log - || die "differences for test_finalpath_cfg.py --endpath --task" $?
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --endpath --task >& finalpath.log || die "failed test_finalpath_cfg.py --endpath --task --schedule" $?
```
Comment (makortel) [inline FWCore/Integration/test/test_finalpath.sh:54]: Should this have
```suggestion
cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --endpath --task --schedule >& finalpath.log || die "failed test_finalpath_cfg.py --endpath --task --schedule" $?
```
?

File: FWCore/Integration/test/test_finalpath.sh [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+grep "thing '.*' TEST" finalpath.log | diff finalpath_expected_found.log - || die "differences for test_finalpath_cfg.py --endpath --task --schedule" $?
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --task >& finalpath.log || die "failed test_finalpath_cfg.py --path --task" $?
+grep "thing '.*' TEST" finalpath.log | diff finalpath_expected_found.log - || die "differences for test_finalpath_cfg.py --path --task" $?
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --task >& finalpath.log || die "failed test_finalpath_cfg.py --path --task --schedule" $?
```
Comment (makortel) [inline FWCore/Integration/test/test_finalpath.sh:60]: Should this have
```suggestion
cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --task --schedule >& finalpath.log || die "failed test_finalpath_cfg.py --path --task --schedule" $?
```
?

File: FWCore/Integration/test/test_finalpath.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+did not find thing '' TEST
+found thing '' TEST
+did not find thing '' TEST
+found thing 'beginLumi' TEST
+found thing 'endLumi' TEST
+found thing 'beginRun' TEST
+found thing 'endRun' TEST
+EOF
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --task >& finalpath.log || die "failed test_finalpath_cfg.py --path --filter" $?
```
Comment (makortel) [inline FWCore/Integration/test/test_finalpath.sh:74]: Should this have
```suggestion
cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --filter >& finalpath.log || die "failed test_finalpath_cfg.py --path --filter" $?
```
or
```suggestion
cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --task >& finalpath.log || die "failed test_finalpath_cfg.py --path --task" $?
```
?

File: FWCore/Integration/test/test_finalpath.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+grep "thing '.*' TEST" finalpath.log | diff finalpath_expected_found.log - || die "differences for test_finalpath_cfg.py --path --filter" $?
+
+
+cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --task >& finalpath.log || die "failed test_finalpath_cfg.py --path --filter --task" $?
```
Comment (makortel) [inline FWCore/Integration/test/test_finalpath.sh:78]: Should this have
```suggestion
cmsRun  ${LOCAL_TEST_DIR}/test_finalpath_cfg.py -- --path --filter --task >& finalpath.log || die "failed test_finalpath_cfg.py --path --filter --task" $?
```
?

File: FWCore/ParameterSet/python/Config.py [hunk 6]
```diff
@@ -1035,6 +1057,7 @@ def splitPython(self, options = PrintOptions()):
         parts.update(self._splitPythonList('sequences', self._itemsInDependencyOrder(self.sequences), options))
         parts.update(self._splitPythonList('paths', self.paths_(), options))
         parts.update(self._splitPythonList('paths', self.endpaths_(), options))
+        parts.update(self._splitPythonList('paths', self.finalaths_(), options))
```
Comment (makortel) [inline FWCore/ParameterSet/python/Config.py:1060]: Should this be
```suggestion
        parts.update(self._splitPythonList('paths', self.finalpaths_(), options))
```
?

File: FWCore/ParameterSet/python/Config.py [hunk 7]
```diff
@@ -1156,19 +1182,49 @@ def _insertPaths(self, processPSet, nodeVisitor):
...
             for path in self.schedule_():
                 pathname = path.label_()
-                scheduledPaths.append(pathname)
                 if pathname in self.endpaths_():
                     endpaths.append(pathname)
+                    scheduledPaths.append(pathname)
+                elif pathname in self.finalpaths_():
+                    finalpaths.append(pathname)
+                    if 1 == len(finalpaths):
+                      scheduledPaths.append("@finalPath")
```
Comment (makortel) [inline FWCore/ParameterSet/python/Config.py:1196]: This looks like duplication of
```py
        endPathWithFinalPathModulesName ="@finalPath"
        ...
        if finalpaths:
          ...
          scheduledPaths.append(endPathWithFinalPathModulesName)
```
below.

File: FWCore/ParameterSet/python/Config.py [hunk 8]
```diff
@@ -2635,6 +2702,36 @@ def testPath(self):
...
+            path += p.c
+            self.assertEqual(str(path),'a+b+c')
+            path = FinalPath(p.a*p.b+p.c)
+            self.assertEqual(str(path),'a+b+c')
+            path = FinalPath(p.a+ p.b*p.c)
+            self.assertEqual(str(path),'a+b+c')
+            path = FinalPath(p.a*(p.b+p.c))
+            self.assertEqual(str(path),'a+b+c')
+            p.es = ESProducer("AnESProducer")
+            self.assertRaises(TypeError,Path,p.es)
```
Comment (makortel) [inline FWCore/ParameterSet/python/Config.py:2721]: Should this test
```suggestion
            self.assertRaises(TypeError,FinalPath,p.es)
```
instead?

File: FWCore/Integration/test/test_finalpath.sh (hunk 1 above)
Comment (Dr15Jones) [reply FWCore/Integration/test/test_finalpath.sh]: fixed

File: FWCore/Integration/test/test_finalpath.sh (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Integration/test/test_finalpath.sh]: fixed

File: FWCore/Integration/test/test_finalpath.sh (hunk 3 above)
Comment (Dr15Jones) [reply FWCore/Integration/test/test_finalpath.sh]: fixed

File: FWCore/Integration/test/test_finalpath.sh (hunk 5 above)
Comment (Dr15Jones) [reply FWCore/Integration/test/test_finalpath.sh]: fixed

File: FWCore/ParameterSet/python/Config.py (hunk 7 above)
Comment (Dr15Jones) [reply FWCore/ParameterSet/python/Config.py]: fixed

File: FWCore/ParameterSet/python/Config.py (hunk 8 above)
Comment (Dr15Jones) [reply FWCore/ParameterSet/python/Config.py]: fixed

File: FWCore/ParameterSet/python/Config.py (hunk 6 above)
Comment (Dr15Jones) [reply FWCore/ParameterSet/python/Config.py]: fixed

File: FWCore/Integration/test/test_finalpath.sh (hunk 4 above)
Comment (Dr15Jones) [reply FWCore/Integration/test/test_finalpath.sh]: fixed

**Comment by Dr15Jones** [comment]
please test

**Comment by fwyzard** [comment]
Once the PR passes all tests, could the changes be backported to `CMSSW_12_2_X` ?

**Comment by makortel** [comment]
> Once the PR passes all tests, could the changes be backported to `CMSSW_12_2_X` ?

Yes, that should be feasible (and was part of the plan, even though I forgot to reply myself in https://github.com/cms-sw/cmssw/issues/36138#issuecomment-975657718).

**Comment by makortel** [comment]
@cmsbuild, please abort

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
+1
- It works, apparently...

**Pull request merged**